File: 2787-ways-to-express-an-integer-as-sum-of-powers/2787-ways-to-express-an-integer-as-sum-of-powers.cpp
class Solution {
public:
    int mod = 1000000007;
    int temp[301][301];
    long long powerLL(int base, int exp, int limit) {
        long long result = 1;
        for (int i = 0; i < exp; i++) {
            result *= base;
            if (result > limit) return result; // stop early to prevent overflow
        }
        return result;
    }

    int dp(int num, int k, int x) {
        if (num == 0) return 1;
        if (num < 0 || k > 300) return 0;

        long long p = powerLL(k, x, num);
        if (p > num) return 0;

        if (temp[num][k] != -1) return temp[num][k];

        int take = dp(num - (int)p, k + 1, x) % mod;
        int not_take = dp(num, k + 1, x) % mod;

        return temp[num][k] = (take + not_take)%mod;

    }
    int numberOfWays(int n, int x) {
        memset(temp, -1, sizeof(temp));
        return dp(n, 1, x);
    }
};

File: 2787-ways-to-express-an-integer-as-sum-of-powers/2787-ways-to-express-an-integer-as-sum-of-powers.py
class Solution:
    def numberOfWays(self, n: int, x: int) -> int:
        mod = 1000000007

        @lru_cache(None)
        def dp(num, k):
            take = 0
            not_take = 0
            if num == 0:
                return 1
            if num < 0 or k**x > num:
                return 0
            take += dp(num-(k**x), k+1)%mod
            not_take += dp(num, k+1)%mod
            return (take + not_take)%mod

        return dp(n, 1)